### PR TITLE
Add support for basic "advanced search"

### PR DESCRIPTION
**Pull request by icewind1991**
Allows for the following searches:

- Whether or not the `read`, `unread` and `answered` flags are set (`is:unread`)
- Search in `from`, `to`, `cc`, `bcc` and `subject` (`from:somebody@example.com`)

Born from the need to find that one unread message in my inbox

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Awesome feature and perfect timing. I was thinking about a possible implementation this week :rocket: 

One of the issues I reflected about was whether we should translate the filters. But I guess that gets unnecessarily complex and those filters are only used by power users anyway, so fine by me to have it in English only.

The second issue I see from the diff is that it won't handle terms that aren't a single word, like `subject: "work report"`, right? That would require a slightly more complex parser, I guess.

Thanks a lot!

File: lib/SearchHelper.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		}
+		$tokens = explode(' ', $filter);
+		$textTokens = [];
+		foreach ($tokens as $token) {
+			if (!$this->parseFilterToken($query, $token)) {
+				$textTokens[] = $token;
+			}
+		}
+		if (count($textTokens)) {
+			$query->text(implode(' ', $textTokens), false);
```
Comment (ChristophWurst) [inline lib/SearchHelper.php]: Could you please share some insight on why you're adding the tokens as text search query here as well, e.g. when they are header text like subjects?

**Comment by stefan-niedermann** [comment]
@ChristophWurst

have you thought (in long term) about graphical filters, e. g. like atlassian jira does?

https://confluence.atlassian.com/jira064/files/720416597/720416653/1/1366249159018/jira-issuenavigator.png

**Comment by ChristophWurst** [comment]
> have you thought (in long term) about graphical filters, e. g. like atlassian jira does?

Are those stored search queries that appear like folders? If so, that would make an awesome feature! Would you mind creating a feature request ticket for that?

File: lib/SearchHelper.php (hunk 1 above)
Comment (icewind1991) [reply lib/SearchHelper.php]: `$textTokens` are all parts of the search query which were not parsed to have a special meaning, and just get used for "normal" text search

**Comment by icewind1991** [comment]
> The second issue I see from the diff is that it won't handle terms that aren't a single word, like subject: "work report", right? That would require a slightly more complex parser, I guess.

That would make the parser significantly more complex yes

> Are those stored search queries that appear like folders

Though about that as well, would be really useful to have

File: lib/SearchHelper.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/SearchHelper.php]: You're right, I misread the condition :see_no_evil:

**Pull request merged**